update pyproject.toml for `ruff format`

tgamblin opened the pull request:
Add ruff configuration to `pyproject.toml`.

This allows `ruff format` in the Spack repository to format all the files we care about, with our line length of 99, the exceptions we already put in place, and excluding things we don't auto-format, like vendored dependencies.

Right now it'll reformat 175 or so files, but only slightly, in places where `ruff` differs from `black`. For the most part I like the ruff format decisions better than `black`, but none of the changes seem too severe.

This does not change `spack style` -- I figure that can come later but this at least will let people start playing with `ruff`.

tgamblin committed:
update pyproject.toml for `ruff format`

Add ruff configuration to `pyproject.toml`.

This allows `ruff format` in the Spack repository to format all
the files we care about, with our line length of 99, the exceptions
we already put in place, and excluding things we don't auto-format,
like vendored dependencies.

Right now it'll reformat 175 or so files, but only slightly, in
places where `ruff` differs from `black`. For the most part I
like the ruff format decisions better than `black`, but none
of the changes seem too severe.

This does not change `spack style` -- I figure that can come
later but this at least will let people start playing with `ruff`.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

adamjstewart commented on pyproject.toml:
```diff
@@ -67,6 +67,21 @@ features = [
+# smallest version ruff seems to support
+line-length = 99
+target-version = "py37"
+include = ["pyproject.toml", "lib/spack/**/*.py", "var/spack/repos/**/*.py", "bin/spack"]
+exclude = ["lib/spack/external"]
```
I would only ever use `extend-include` and `extend-exclude`, there are a lot of defaults we don't want to override such as excluding `.git`. No use including pyproject.toml, it can't be formatted anyway.

adamjstewart commented on pyproject.toml:
```diff
@@ -67,6 +67,21 @@ features = [
+[tool.ruff]
+# smallest version ruff seems to support
+line-length = 99
+target-version = "py37"
```
We don't normally use this, we normally use:
```toml
[project]
requires-python = ">=3.8"
```
I'm guessing this is only needed because we still support 3.6? Wonder if we could drop that soon.

adamjstewart commented on pyproject.toml:
```diff
@@ -67,6 +67,21 @@ features = [
 ]
 
+[tool.ruff]
+# smallest version ruff seems to support
```
I'm guessing this comment should be one line below?

adamjstewart commented on pyproject.toml:
```diff
@@ -67,6 +67,21 @@ features = [
+extend-exclude = ["*.pyi"]
+
+[tool.ruff.lint]
+ignore = ["E731", "E203"]
```
At a bare minimum I would `extend-select = ["I"]` so we can also get rid of isort, I think `UP` and `RUF` are also quite useful.

adamjstewart reviewed (commented):
See https://github.com/microsoft/torchgeo/blob/main/pyproject.toml for the settings I use in my other project. Might be worth playing around with other rules to see which ones solve issues and which ones are just a headache.

tgamblin committed:
changes recommended by Adam

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
need F811 for package.py `@when` functions

tgamblin committed:
add ruff isort settings

tgamblin commented:
@adamjstewart I added all your suggestions (which also simplified `black` config), except the ones that would go above and beyond what we currently do. I also mapped our `isort` rules to `ruff`' config syntax and submitted a few things `ruff` catches that also pass `spack style` in #48831 and #48832.

I think this is ready to go.

adamjstewart commented on pyproject.toml:
```diff
@@ -67,9 +68,42 @@ features = [
 
+[tool.ruff]
+line-length = 99
+extend-include = ["pyproject.toml", "lib/spack/**/*.py", "var/spack/repos/**/*.py", "bin/spack"]
```
pyproject.toml can't be formatted. All .py files are included by default. I think you only need to add bin/spack.

tgamblin replied:
done

tgamblin committed:
ruff: No need to include pyproject.toml or .py files in include

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>
